keps/sig-network/20191205-kube-proxy-IP-node-binding.md

```diff
@@ -0,0 +1,70 @@
+---
+title: Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+authors:
+  - "@Sh4d1"
+owning-sig: sig-network
+participating-sigs:
+  - sig-cloudprovider
+reviewers:
+  - "@thockin"
+  - "@cheftako"
+  - TBD
+approvers:
+  - "@thockin"
+  - TBD
+editor: TBD
+creation-date: 2019-12-05
+last-updated: 2019-12-05
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Give the ability to the LoadBalancer provisioner to disable the LB IP binding
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+<!-- /toc -->
+
+## Summary
+
+The different kube proxy implentations (at least ipvs and iptables) are today sort of binding the External IPs of the LoadBalancer Service to each node. (iptables is creating some rules to redirect directly the packets to the service and ipvs is binding the IP to one interface on the node). This feature is exisiting because:
+
+1. To redirect pod traffic that is intended for LoadBalancer IP directly to the service, bypassing the loadbalancer for path optimization.
+2. Some LBs are sending packet with the destination IP equal to the LB IP, so the IP needs to be routable directly to the right service (otherwise looping back and forth)
+3. (TBD)
```

Comment:
3. Provide means of load-balancing external traffic in environments without a Kubernetes-aware external load-balancer i.e. private-clouds.

Reply:
For me, as mainly a private-cloud user, this is the main reason.

Reply:
Lars, can you say more?

I kind of want to argue that the current behavior is *always* wrong, but your comment here gives me pause.

Even so, changing it in all cases will be hard - it will require on-node changes on some providers.  We can tell if traffic came from "local" or not, so we should be able to only leave this rule in place for traffic through LBs.  I'd like to see us explore that, but I ACK that it is bigger than this KEP.

@bowei I think GCP is one of the few who require it.

Reply:
@thockin If there are no K8s aware external load-balancer the NodePorts become useless, but load-balancing based on external address is crucial. IMHO it is wrong to see load-balancing of external addresses as a (flawed) help-function for cloud providers who _have_ an external load-balancer. Load-balancing on external address is most likely the *only* option for private clouds. IMO this KEP should be extended to also provide the option to completely disable NodePorts (with it's security issues and sync overhead) in private clusters that relies only on load-balancing on external address. Please take a broader view and do not only concentrate on problems for cloud-providers.

Reply:
@uablrek FYI there's a KEP for disabling node ports here https://github.com/kubernetes/enhancements/pull/1514 that covers some of the issues regarding node ports you mentioned. PTAL, I would appreciate your feedback on the motivation / user story section.

Reply:
Thanks for the pointer. I will check it tomorrow.